pallets/gated-marketplace/src/functions.rs

```diff
@@ -150,9 +151,328 @@ impl<T: Config> Pallet<T> {
         Ok(())
     }
 
+    pub fn do_enlist_sell_offer(authority: T::AccountId, marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId, price: BalanceOf<T>,) -> DispatchResult {
+        //This function is only called by the owner of the marketplace
+        //ensure the marketplace exists
+        ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+
+        //ensure the collection exists
+        if let Some(a) = pallet_uniques::Pallet::<T>::owner(collection_id, item_id) {
+            ensure!(a == authority, Error::<T>::NotOwner);
+        } else {
+            Err(Error::<T>::CollectionNotFound)?;
+        }
+
+        //ensure the price is valid
+        Self::is_the_price_valid(price)?;
+
+        //Add timestamp to the offer
+        let(timestamp, timestamp2) = Self::get_timestamp_in_milliseconds().ok_or(Error::<T>::TimestampError)?;
+
+        //create an offer_id
+        //TODO: create an offer id generator, used in cases where the offer_id generated is not unique
+        let offer_id = (marketplace_id, authority.clone(), collection_id, timestamp, timestamp2).using_encoded(blake2_256);
+
+        //create offer structure 
+        let offer_data = OfferData::<T> {
+            marketplace_id: marketplace_id,
+            collection_id: collection_id,
+            item_id: item_id,
+            creator: authority.clone(),
+            price: price,
+            creation_date: timestamp,
+            expiration_date: timestamp2,
+            status: OfferStatus::Open,
+            offer_type: OfferType::SellOrder,
+            buyer: None,
+        };
+
+
+        //ensure there is no a previous sell offer for this item
+        Self::is_item_already_for_sale(collection_id, item_id, marketplace_id)?;
+        
+        //insert in OffersByItem
+        <OffersByItem<T>>::try_mutate(collection_id, item_id, |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        //insert in OffersByAccount
+        <OffersByAccount<T>>::try_mutate(authority.clone(), |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        //insert in OffersInfo
+        // ensure the offer_id doesn't exist
+        ensure!(!<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferAlreadyExists);
+        <OffersInfo<T>>::insert(offer_id, offer_data);
+
+        //Insert in OffersByMarketplace
+        <OffersByMarketplace<T>>::try_mutate(marketplace_id, |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        Self::deposit_event(Event::OfferStored(collection_id, item_id));
+        Ok(())
+    }
+
+    pub fn do_enlist_buy_offer(authority: T::AccountId, marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId, price: BalanceOf<T>,) -> DispatchResult {
+        //TODO: ensure the user is a Marketparticipant
```

Comment:
Adding this functionality on RBAC branch

Reply:
Ok.